### ISSUE TITLE
QueryBuilder have IN lists query and append script issue

### ISSUE THREAD

**Issue by FreshmanChina** [Drizin/DapperQueryBuilder#22]
Hello,I have  a script replace by  6 length c#  list  var through by In lists query,then apply append script,the parameter have wrong place in action script,it overlap with other parameter !Please forgive me my poor English!

**Comment by Drizin** [comment]
Can you share the code?

**Comment by Drizin** [comment]
@FreshmanChina can you share the code?

**Comment by FreshmanChina** [comment]
`using DapperQueryBuilder;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;

namespace ConsoleApp7
{
    class Program
    {
        static void Main(string[] args)
        {
            using SqlConnection conn = new SqlConnection("Data Source=(localdb)\\mssqllocaldb;Initial Catalog=testdb;Integrated Security=True;");
            string v = "a";
            List<int> numList = new List<int> { 1, 2, 3, 4, 5, 6 };

            FormattableString script = @$"
declare @v1 nvarchar(10)={v}
declare @v2 nvarchar(10)={v}
select 1 from tb where name in {numList}
declare @v3 nvarchar(10)={v}
declare @v4 nvarchar(10)={v}
declare @v5 nvarchar(10)={v}
declare @v6 nvarchar(10)={v}
declare @v7 nvarchar(10)={v}
declare @v8 nvarchar(10)={v}
declare @v9 nvarchar(10)={v}
declare @v10 nvarchar(10)={v}
declare @v11 nvarchar(10)={v}
declare @v12 nvarchar(10)={v}
declare @v13 nvarchar(10)={v}
declare @v14 nvarchar(10)={v}
declare @v15 nvarchar(10)={v}
declare @v16 nvarchar(10)={v}
declare @v17 nvarchar(10)={v}
declare @v18 nvarchar(10)={v}
declare @v19 nvarchar(10)={v}
declare @v20 nvarchar(10)={v}
declare @v21 nvarchar(10)={v}
declare @v22 nvarchar(10)={v}
declare @v23 nvarchar(10)={v}
select 'ok'
";
            
            QueryBuilder queryBuilder = conn.QueryBuilder(script);
            string rst = queryBuilder.QueryFirstOrDefault<string>();
            Console.WriteLine(rst);
        }
    }
}
`

**Comment by FreshmanChina** [comment]
![image](https://user-images.githubusercontent.com/16652353/119697888-41de1a00-be83-11eb-9ea3-68e1059b02cb.png)

**Comment by FreshmanChina** [comment]
exec sp_executesql N'declare @v1 nvarchar(10)=@p0
declare @v2 nvarchar(10)=@p1
select 1 from tb where name in (@p21,@p22,@p23,@p24,@p25,@p26)
declare @v3 nvarchar(10)=@p3
declare @v4 nvarchar(10)=@p4
declare @v5 nvarchar(10)=@p5
declare @v6 nvarchar(10)=@p6
declare @v7 nvarchar(10)=@p7
declare @v8 nvarchar(10)=@p8
declare @v9 nvarchar(10)=@p9
declare @v10 nvarchar(10)=@p10
declare @v11 nvarchar(10)=@p11
declare @v12 nvarchar(10)=@p12
declare @v13 nvarchar(10)=@p13
declare @v14 nvarchar(10)=@p14
declare @v15 nvarchar(10)=@p15
declare @v16 nvarchar(10)=@p16
declare @v17 nvarchar(10)=@p17
declare @v18 nvarchar(10)=@p18
declare @v19 nvarchar(10)=@p19
declare @v20 nvarchar(10)=@p20
declare @v21 nvarchar(10)=@p21
declare @v22 nvarchar(10)=@p22
declare @v23 nvarchar(10)=@p23
select ''ok''
',N'@p0 nvarchar(4000),@p1 nvarchar(4000),@p21 nvarchar(4000),@p22 nvarchar(4000),@p23 nvarchar(4000),@p24 int,@p25 int,@p26 int,@p3 nvarchar(4000),@p4 nvarchar(4000),@p5 nvarchar(4000),@p6 nvarchar(4000),@p7 nvarchar(4000),@p8 nvarchar(4000),@p9 nvarchar(4000),@p10 nvarchar(4000),@p11 nvarchar(4000),@p12 nvarchar(4000),@p13 nvarchar(4000),@p14 nvarchar(4000),@p15 nvarchar(4000),@p16 nvarchar(4000),@p17 nvarchar(4000),@p18 nvarchar(4000),@p19 nvarchar(4000),@p20 nvarchar(4000)',@p0=N'a',@p1=N'a',@p21=N'a',@p22=N'a',@p23=N'a',@p24=4,@p25=5,@p26=6,@p3=N'a',@p4=N'a',@p5=N'a',@p6=N'a',@p7=N'a',@p8=N'a',@p9=N'a',@p10=N'a',@p11=N'a',@p12=N'a',@p13=N'a',@p14=N'a',@p15=N'a',@p16=N'a',@p17=N'a',@p18=N'a',@p19=N'a',@p20=N'a'

**Comment by FreshmanChina** [comment]
Hello,please pay attention to the p21,p22,p23,p24,p25,p26,it should to 1,2,3,4,5,6,but I monitor the run script the p21='a',p22='a',p23='a',p24=4,p25=5,p26=6

**Comment by FreshmanChina** [comment]
Hello,I reply from here:
https://github.com/Drizin/DapperQueryBuilder/issues/22
thank you!


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Drizin/DapperQueryBuilder"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年5月26日(星期三) 晚上9:38
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Drizin/DapperQueryBuilder] QueryBuilder IN lists query parameter  overlap with other parameter (#22)





 
@FreshmanChina can you share the code?

**Comment by FreshmanChina** [comment]
I use latest version:1.1.0

**Comment by Drizin** [comment]
@FreshmanChina - I haven't checked deeply, but it looks like a bug in Dapper.
You can probably reproduce the same issue using pure Dapper. Can you try it?

**Comment by Drizin** [comment]
Looks like it's the same as [reported here](https://github.com/DapperLib/Dapper/issues/898) ?

**Comment by FreshmanChina** [comment]
Yes, it's the same as reported above,Actually, Setting the in clause paramter name with random name instead of the static prefix name,will fix the issue.

**Comment by Drizin** [comment]
I didn't get it. Can you share how you fixed it?

It looks like if `@p1` is an array Dapper will automatically expand it into `@p11`, `@p12`, etc. And in this case if you have more than 12 elements (`@p11`) it will collide names. 

I think the easiest solution here would be using fixed 3-digits length for parameters. Like `@p001`, which in this case if it was an array it would be expanded to `@p0011` and would not collide with the 12-th parameter `@p011`. Problem would only occur if `@p101` was also an array and in case could collide with 1012-th parameter `@p1011` (which is very unlikely).

For an immediate fix you can try a workaround to force your array variable to be 2 digits:
```
declare @dummy0 int={0} -- this will be @p0
declare @dummy1 int={0} -- ...
declare @dummy2 int={0}
declare @dummy3 int={0}
declare @dummy4 int={0}
declare @dummy5 int={0}
declare @dummy6 int={0}
declare @dummy7 int={0}
declare @dummy8 int={0}
declare @dummy9 int={0} -- this will be @p9

-- now you can use your array:
select 1 from tb where name in {numList}
--  because it will be named as @p10 
-- and if it's an array Dapper will expand it into @p101, @p102, etc
-- so that does not conflict with your parameters unless you have 102+ parameters.
```

**Comment by FreshmanChina** [comment]
Thank you.This is a good idea!I have a idea,but I don't know if it will work,name the parameter with guid string?
![image](https://user-images.githubusercontent.com/16652353/119913874-9f608c80-bf91-11eb-90d3-d004a838e971.png)

**Comment by Drizin** [comment]
@FreshmanChina  I don't like the idea of random names because it would require changes in all unit tests, and also would make debugging more difficult. But I have made a fix - would you like to test and let me know if it works? Thanks!
